### DISCUSSION TITLE
[Angular + tailwind + postcss] warn - The `content` option in your Tailwind CSS configuration is missing or empty.

### DISCUSSION THREAD

**Discussion by tom-vism** [tailwindlabs/tailwindcss#13046] [Help]
I use NX monorepo with 17.2.2 Angular version.
And latest tailwindcss 3 and postcss 8

This is a monorepo, and an app has a tailwind.config.js and postcss.config.json file in it's project root.

When I include "tailwindcss" in postcss configuration file I get an error when I serve Angular app (using the recent "application" builder), the compiler shows some warnings (about "content" property not being found) and then errors because it ignores (does not detect) tailwind configuration in the project root. But it works if I put tailwind.config.js in the workspace root.

Is this a tailwind plugin problem and how can I locate the tailwind.config.file?

The warnings and errors:

![image](https://github.com/tailwindlabs/tailwindcss/assets/142388200/751e9ffa-32d7-4711-9b96-29cf1cb9a3a3)

**Minimal Reproduction**
1. Create an NX project for Angular ^17.2 with the workspace structure like below (reference to the older issue: https://github.com/angular/angular-cli/issues/27005.)
2. Use the application builder and make necessary changes in project.json of the app (known as angular.json)
3. Create the postcss configuration file (postcss.config.json) in the project root

My workspace structure looks something like this:

```
my-nx-workspace/
|-- apps/
|   |-- group
|   |   |-- app1/
|   |   |   |-- src/
|   |   |   |-- postcss.config.json
|   |   |   |-- tailwind.config.js
|   |   |   |-- project.json
|   |   |   |-- ...
|   |   |-- app2/
|   |   |   |-- src/
|-- libs/
|   |-- shared-lib/
|   |   |-- src/
|-- tools/
|-- node_modules/
|-- nx.json
|-- package.json
|-- tsconfig.base.json
|-- tailwind-base-configuration.js

```

My postcss.config.json
```
{
  "plugins": {
    "postcss-import": {},
    "tailwindcss": {},
    "autoprefixer": {}
  }
}
```
My project tailwind.config.js looks like this

```
const { createGlobPatternsForDependencies } = require('@nx/angular/tailwind');
const { join } = require('path');
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    join(__dirname, 'src/**/!(*.stories|*.spec).{ts,html}'),
    ...createGlobPatternsForDependencies(__dirname),
  ],
  presets: [require('../../../tailwind-base-configuration.js')],
  theme: {
  }
};
```

**Comment by alex-krasikau** [comment]
Hey @tom-vism!

It seems that your application doesn't recognize the Tailwind config at all. Let's assume you're running the `app1` application. Based on the shared project structure, I can see that both the Tailwind and PostCSS configs are located outside of the project folder, making them invisible to your project build tool.

If this is the case, you can move and copy the same PostCSS and Tailwind CSS configs into both the `app1` and `app2` folders. This will allow them to pick up the configuration correctly.

**Reply by tom-vism** [reply]
I updated the workspace structure in the description, I drew it incorrectly. My configuration files are inside app1 on the same level as src folder (just as you described), but still getting errors.

**Reply by alex-krasikau** [reply]
Alright, let's continue

1. Can you confirm if the styling still work while you see this warning? I searched for some related issues and found some people who have complained on having this warning while the Tailwind CSS still did it thing. At the end of the day the warning can be ignored.

This will also help us understand if the tailwind config is picked up by the project, so we can continue investigation.

2. Can you please try renaming your`tailwind-base-configuration.js` to `tailwind.config.js` and check if you still observe the warning.

3. A discussion similar to yours was held on NX with a thread that might help you find a solution. Check it out - https://github.com/tailwindlabs/tailwindcss/discussions/7695#discussioncomment-3162111.

4. As you have a pretty complex setup, it would be great if you can create a reproduction repo and reproduction steps, so we can dig into it.

**Reply by tom-vism** [reply]
1. styles are completely broken, tailwind default classes not applied.
2. When I rename the file in the workspace root, the warning is there and styles are not applied, but it's expected since it has no "content" property. When I add the "content" property and point to apps/\*\*... and libs/\*\*... then tailwind starts work (updated)
However, the tailwind config file in the project root still seems to be completely ignored.

**Reply by alex-krasikau** [reply]
That's the thing, your project picks up and expects the tailwind config present at the root of the workspace. I lack sufficient experience to offer valuable advice. I suggest deleting the current postcss and Tailwind files and setting up from scratch. Sometimes, simple errors made during setup can be challenging to debug later on.

https://nx.dev/recipes/angular/using-tailwind-css-with-angular-projects

**Reply by tom-vism** [reply]
The thing it worked everything before in Angular 17.1, but in Angular 17.2 new "application" builder allows to use postcss configuration and this where problems start.
I'll try to create minimal reproduction repo (from scratch)

**Reply by tom-vism** [reply]
I have created an example monorepo. I followed the guide you gave from scratch to set up the tailwind. As per documentation, the shared tailwind config file is put in tailwind-preset folder in libs, and then app1 and shared-lib have their own tailwind.config.js file (yes even shared-lib because buildable libraries must have their own tailwind.config.js file).
Everything works fine until I needed to create postcss configuration file. I created it in json format because otherwise Angular not picking it up. It has 3 plugins, the first one (postcss-import) is needed to be able to import headers.css in styles.css.

https://github.com/tom-vism/angular-monorepo

**Reply by tom-vism** [reply]
To run the project use nx serve app1. 
To make the tailwind work without postcss, you'd need to remove postcss.config.json and comment the '.../headers.css' import statement in styles.css of the app1.

**Reply by tom-vism** [reply]
up

**Comment by LeonasReckus** [comment]
Up
